Pass through piggy-bank example of smart contract tutorial (#65)
Concordium/concordium.github.io: source/smart-contracts/tutorials/piggy-bank/writing.rst

===================

The contract we are going to build in this tutorial is going to act as a classic

> piggy bank. Everyone should be able to put money into it, but only the owner

Reviewer: I think we need to use GTU and not money.

We are specifically using the dedicated amount to contribute and using the chain to keep track of it, and the chain keeps track of GTU only.

Author: ```suggestion
piggy bank. Everyone should be able to put money in the form of GTU into it, but only the owner
```
Sounds good to me--I'd like to use money in the first use of this because a piggy bank usually contains money so I want to use that word just once, and below I'll say GTU.

Reviewer: Ok.

can smash it and retrieve the GTU inside. Once the piggy bank has been

> smashed, it should not be possible to add money to it.

Reviewer: ```suggestion
smashed, it should not be possible to add GTU to it.
```

.. todo::

   Add image of piggy bank.

The piggy-bank smart contract is going to contain a function for setting up a
new piggy bank and two functions for updating a piggy bank; one is for everyone
to use for inserting GTU, the other is for the owner to smash the piggy bank and
prevent further interaction.

Specifying the state
====================

> To implement a piggy bank we need to keep track of the amount of money it holds,

Reviewer: ```suggestion
To implement a piggy bank we need to keep track of the amount of GTU it holds,
```

Define interaction with piggy banks
===================================

We have now defined how instances of our smart contract are created, and the
smart contract is in principle a valid contract.
However, we would also like to define how to interact with instances of our
contract.

> Specifically how to add money to it and how to smash a piggy bank.

Reviewer: ```suggestion
Specifically how to add GTU to it and how to smash a piggy bank.
```

      todo!("Implement")
   }

The ``contract`` attribute supplies the name of the contract to the macro.
This name should match the name in the corresponding attribute in |init|_
(``"PiggyBank"`` in our case). It also requires a name to identify this
particular receive function using ``name``. The name and contract attributes
each have to be unique within a smart contract module.

The return type of the function is ``ReceiveResult<A>``, which is an alias for
``Result<A, Reject>``.
Here, ``A`` implements |HasActions|, which exposes functions for creating
various :ref:`actions<action-descriptions>`.

In this contract we will use the **Accept** and **Simple Transfer** actions.

Inserting GTU
-------------

> The first interaction we will specify for our piggy bank is how to put money into it.

Reviewer 2: ```suggestion
The first interaction we will specify for our piggy bank is how to insert GTU into it.
```
Following the assumption that we should say GTU instead of `money`

       Ok(A::simple_transfer(&owner, balance))
   }

.. .. note::
   Since a blockchain is a decentralized system, one might think we have to
   worry about the usual problems when dealing with mutable state. Problems
   such as race conditions, but the semantics of smart contracts require the
   execution to be atomic, in order to reach consensus.

We now have all the parts for our piggy bank smart contract. Before we move on
to testing it, we check that it builds by running:

.. code-block:: console

   $cargo concordium build

> Thich should succeed if everything is set up correctly. Otherwise, compare your

Reviewer: ```suggestion
This should succeed if everything is set up correctly. Otherwise, compare your
```
